Updated version to latest 1.4.5

o-mdr opened the pull request:
Added support for v1.4.5 zstd.

This PR also contains:
- Implementation to support v1.4.5 of [zstd](https://github.com/facebook/zstd/releases/tag/v1.4.5)
- Meson patch definitions, taken from https://github.com/facebook/zstd/tree/dev/build/meson with minor modifications (hardcoded version, updated paths to subprojects)

Tested on Ubuntu 20 (g++) and Mac OS 10.14.6 (clang), but this doesn't work on Windows with MSVC v19 compiler. Windows build with MSVC is failing on `zstd` repo itself, I have [raised an issue](https://github.com/facebook/zstd/issues/2358) and there is a [PR that removes support for MSVC 19 compiler](https://github.com/facebook/zstd/pull/2363)

Testing:
1. Download https://github.com/facebook/zstd/releases/tag/v1.4.5 released sources
2. Decompress into a folder
3. Copy `meson.build` and `meson_options.txt` from this PR into the root source folder of zstd
4. Run tests, for example

```
cd zstd
meson setup builddir
meson compile -C builddir
meson configure builddir -Dbin_tests=true
# Long running tasks
meson test -C builddir
```

shawnl committed:
1.3.3 (#1) (#2)

avagordon01 committed:
change find_library('m') required to false to build on windows

legeana commented:
I am getting a lot of undefined reference errors, could you please take a look?

```
cc  -o libzstd.so.1.4.5 libzstd.so.1.4.5.p/lib_common_entropy_common.c.o libzstd.so.1.4.5.p/lib_common_fse_decompress.c.o libzstd.so.1.4.5.p/lib_common_threading.c.o libzstd.so.1.4.5.p/lib_common_pool.c.o libzstd.so.1.4.5.p/lib_common_zstd_common.c.o libzstd.so.1.4.5.p/lib_common_error_private.c.o libzstd.so.1.4.5.p/lib_common_xxhash.c.o libzstd.so.1.4.5.p/lib_compress_fse_compress.c.o libzstd.so.1.4.5.p/lib_compress_huf_compress.c.o libzstd.so.1.4.5.p/lib_compress_zstd_compress.c.o libzstd.so.1.4.5.p/lib_compress_zstd_fast.c.o libzstd.so.1.4.5.p/lib_compress_zstd_double_fast.c.o libzstd.so.1.4.5.p/lib_compress_zstd_lazy.c.o libzstd.so.1.4.5.p/lib_compress_zstd_opt.c.o libzstd.so.1.4.5.p/lib_compress_zstd_ldm.c.o libzstd.so.1.4.5.p/lib_compress_zstdmt_compress.c.o libzstd.so.1.4.5.p/lib_decompress_huf_decompress.c.o libzstd.so.1.4.5.p/lib_decompress_zstd_decompress.c.o libzstd.so.1.4.5.p/lib_dictBuilder_cover.c.o libzstd.so.1.4.5.p/lib_dictBuilder_divsufsort.c.o libzstd.so.1.4.5.p/lib_dictBuilder_zdict.c.o libzstd.so.1.4.5.p/lib_deprecated_zbuff_common.c.o libzstd.so.1.4.5.p/lib_deprecated_zbuff_compress.c.o libzstd.so.1.4.5.p/lib_deprecated_zbuff_decompress.c.o libzstd.so.1.4.5.p/lib_legacy_zstd_v04.c.o libzstd.so.1.4.5.p/lib_legacy_zstd_v05.c.o libzstd.so.1.4.5.p/lib_legacy_zstd_v06.c.o libzstd.so.1.4.5.p/lib_legacy_zstd_v07.c.o -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,--start-group -Wl,-soname,libzstd.so.1 -Wl,--end-group
/usr/bin/ld: libzstd.so.1.4.5.p/lib_compress_fse_compress.c.o: in function `FSE_compress_wksp':
/tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/fse_compress.c:652: undefined reference to `HIST_count_wksp'
/usr/bin/ld: libzstd.so.1.4.5.p/lib_compress_huf_compress.c.o: in function `HUF_compressWeights':
/tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/huf_compress.c:82: undefined reference to `HIST_count_simple'
/usr/bin/ld: libzstd.so.1.4.5.p/lib_compress_huf_compress.c.o: in function `HUF_compress_internal':
/tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/huf_compress.c:672: undefined reference to `HIST_count_wksp'
/usr/bin/ld: libzstd.so.1.4.5.p/lib_compress_zstd_compress.c.o: in function `ZSTD_compressSequences_internal':
/tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:1996: undefined reference to `ZSTD_compressLiterals'
/usr/bin/ld: /tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2038: undefined reference to `HIST_countFast_wksp'
/usr/bin/ld: /tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2041: undefined reference to `ZSTD_selectEncodingType'
/usr/bin/ld: /tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2048: undefined reference to `ZSTD_buildCTable'
/usr/bin/ld: /tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2064: undefined reference to `HIST_countFast_wksp'
/usr/bin/ld: /tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2070: undefined reference to `ZSTD_selectEncodingType'
/usr/bin/ld: /tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2076: undefined reference to `ZSTD_buildCTable'
/usr/bin/ld: /tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2092: undefined reference to `HIST_countFast_wksp'
/usr/bin/ld: /tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2096: undefined reference to `ZSTD_selectEncodingType'
/usr/bin/ld: /tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2102: undefined reference to `ZSTD_buildCTable'
/usr/bin/ld: /tmp/tmpebgc73ts/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2119: undefined reference to `ZSTD_encodeSequences'
```

o-mdr commented:
I've had some time to look at this and got your issue reproduced. Clearly, it's not something in the environment. 

It appears the build files have been updated quite a bit so that the patch no longer works. I spent a bit of time creating a new patch which I will attach as a zip below. 

Here's an outline of the changes:

### meson.build
User main meson file that depends on `zstd`, notice change in the dependency name ([`libzstd_dep`](https://github.com/facebook/zstd/blob/dev/build/meson/lib/meson.build) is how zstd calls the dependency, it think it's best to keep it in sync)
```
project('foo', 'cpp', default_options : 
    ['cpp_std=c++17', 'buildtype=release'], 
    version : '0.1', license: 'MIT')

subproject ('zstd')

# IMPORTANT: use `libzstd_dep` for the dependency name
zstd = dependency('libzstd_dep', fallback : ['zstd', 'libzstd_dep'])

src = [ ... ]
library(src, dependencies: [zstd])
```

### zstd.wrap 
File located in `subprojects/zstd.wrap`, note the patch URL needs to be updated, obviously.
```
[wrap-file]
directory = zstd-1.4.5

source_url = https://github.com/facebook/zstd/releases/download/v1.4.5/zstd-1.4.5.tar.gz
source_filename = zstd-1.4.5.tar.gz
source_hash = 98e91c7c6bf162bf90e4e70fdbc41a8188b9fa8de5ad840c401198014406ce9e

patch_url = file:///home/om/Downloads/zstd-1.4.5.zip
patch_filename = zstd-1.4.5.zip
patch_hash = 8eac7d6b85b37e3af49aee65179ab966c379626dd3e73c06c8c257c11514e858
```

### Patch files
This is based on https://github.com/facebook/zstd/tree/dev/build/meson, also includes https://github.com/facebook/zstd/tree/dev/build/meson/lib, but not other dependencies, such as `programs`, `tests`, `contrib`, which I have left commented in the patch file. 

I have also hardcoded the version to remove the dependency on python - the version is hardcoded to 1.4.5 in the patch root meson.build file.

Patch file: [zstd-1.4.5.zip](https://github.com/mesonbuild/zstd/files/5354856/zstd-1.4.5.zip)

Please let me know if this works for you. Thanks!

o-mdr commented:
Oh, just noticed someone made a simplified patch file already... Oh, well :D

legeana commented:
Your changes didn't help unfortunately:
```
[29/29] Linking target libzstd.so.1.4.5
FAILED: libzstd.so.1.4.5
cc  -o libzstd.so.1.4.5 libzstd.so.1.4.5.p/lib_common_entropy_common.c.o libzstd.so.1.4.5.p/lib_common_fse_decompress.c.o libzstd.so.1.4.5.p/lib_common_threading.c.o libzstd.so.1.4.5.p/lib_common_pool.c.o libzstd.so.1.4.5.p/lib_common_zstd_common.c.o libzstd.so.1.4.5.p/lib_common_error_private.c.o libzstd.so.1.4.5.p/lib_common_xxhash.c.o libzstd.so.1.4.5.p/lib_compress_fse_compress.c.o libzstd.so.1.4.5.p/lib_compress_huf_compress.c.o libzstd.so.1.4.5.p/lib_compress_zstd_compress.c.o libzstd.so.1.4.5.p/lib_compress_zstd_fast.c.o libzstd.so.1.4.5.p/lib_compress_zstd_double_fast.c.o libzstd.so.1.4.5.p/lib_compress_zstd_lazy.c.o libzstd.so.1.4.5.p/lib_compress_zstd_opt.c.o libzstd.so.1.4.5.p/lib_compress_zstd_ldm.c.o libzstd.so.1.4.5.p/lib_compress_zstdmt_compress.c.o libzstd.so.1.4.5.p/lib_decompress_huf_decompress.c.o libzstd.so.1.4.5.p/lib_decompress_zstd_decompress.c.o libzstd.so.1.4.5.p/lib_dictBuilder_cover.c.o libzstd.so.1.4.5.p/lib_dictBuilder_divsufsort.c.o libzstd.so.1.4.5.p/lib_dictBuilder_zdict.c.o libzstd.so.1.4.5.p/lib_deprecated_zbuff_common.c.o libzstd.so.1.4.5.p/lib_deprecated_zbuff_compress.c.o libzstd.so.1.4.5.p/lib_deprecated_zbuff_decompress.c.o libzstd.so.1.4.5.p/lib_legacy_zstd_v04.c.o libzstd.so.1.4.5.p/lib_legacy_zstd_v05.c.o libzstd.so.1.4.5.p/lib_legacy_zstd_v06.c.o libzstd.so.1.4.5.p/lib_legacy_zstd_v07.c.o -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,--start-group -Wl,-soname,libzstd.so.1 -Wl,--end-group
/usr/bin/ld: libzstd.so.1.4.5.p/lib_compress_fse_compress.c.o: in function `FSE_compress_wksp':
/tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/fse_compress.c:652: undefined reference to `HIST_count_wksp'
/usr/bin/ld: libzstd.so.1.4.5.p/lib_compress_huf_compress.c.o: in function `HUF_compressWeights':
/tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/huf_compress.c:82: undefined reference to `HIST_count_simple'
/usr/bin/ld: libzstd.so.1.4.5.p/lib_compress_huf_compress.c.o: in function `HUF_compress_internal':
/tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/huf_compress.c:672: undefined reference to `HIST_count_wksp'
/usr/bin/ld: libzstd.so.1.4.5.p/lib_compress_zstd_compress.c.o: in function `ZSTD_compressSequences_internal':
/tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:1996: undefined reference to `ZSTD_compressLiterals'
/usr/bin/ld: /tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2038: undefined reference to `HIST_countFast_wksp'
/usr/bin/ld: /tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2041: undefined reference to `ZSTD_selectEncodingType'
/usr/bin/ld: /tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2048: undefined reference to `ZSTD_buildCTable'
/usr/bin/ld: /tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2064: undefined reference to `HIST_countFast_wksp'
/usr/bin/ld: /tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2070: undefined reference to `ZSTD_selectEncodingType'
/usr/bin/ld: /tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2076: undefined reference to `ZSTD_buildCTable'
/usr/bin/ld: /tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2092: undefined reference to `HIST_countFast_wksp'
/usr/bin/ld: /tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2096: undefined reference to `ZSTD_selectEncodingType'
/usr/bin/ld: /tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2102: undefined reference to `ZSTD_buildCTable'
/usr/bin/ld: /tmp/tmpbo6kb8yc/bin/../src/zstd-1.4.5/lib/compress/zstd_compress.c:2119: undefined reference to `ZSTD_encodeSequences'
```

o-mdr commented:
Hi @legeana, that's strange, I am able to compile and use the newer version of zstd.

Is it an open-source project you testing? If you point me to where/how you use it I may be able to spot why it's not working.  Could you please check the following:

- Could you delete `subprojects/packagecache` and `subprojects/zstd-1.4.5`? This got me a few times, seems like meson is caching dependencies and would not re-build by default, even if zstd 1.4.5 has changed. The error message seems to be similar to your previous attempt, I wonder if this is because meson didn't pick up the changes.

- Could you confirm you made the patch and it has the same sha2 to what i have?

- Could you post the relevant part of meson build where it's building the new version of zst lib?

- Could you confirm you used `zstd = dependency('libzstd_dep', fallback : ['zstd', 'libzstd_dep'])` in your project's meson file?

jpakkane commented:
I tried this on macOS and go an equivalent error:

```
Undefined symbols for architecture x86_64:
  "_HIST_countFast_wksp", referenced from:
      _ZSTD_compressSequences_internal in lib_compress_zstd_compress.c.o
  "_HIST_count_simple", referenced from:
      _HUF_compressWeights in lib_compress_huf_compress.c.o
      _ZSTD_rescaleFreqs in lib_compress_zstd_opt.c.o
  "_HIST_count_wksp", referenced from:
      _FSE_compress_wksp in lib_compress_fse_compress.c.o
      _HUF_compress_internal in lib_compress_huf_compress.c.o
  "_ZDICT_optimizeTrainFromBuffer_fastCover", referenced from:
      _ZDICT_trainFromBuffer in lib_dictBuilder_zdict.c.o

etc etc
```

The actual function seems to be defined in `lib/compress/hist.c`, which is not listed in the main Meson file's sources.

While we're at it, there is a lot of pointless path joining happening. Instead of this:

```meson
libzstd_srcs = [
    join_paths(common_dir, 'entropy_common.c'),
...
```

you could do the much simpler version of:

```meson
libzstd_srcs = files(
    'common/entropy_common.c',
...
```

Or, if you absolutely have to use string joining:

```meson
common_dir = 'common'
libzstd_srcs = files(
    common_dir / 'entropy_common.c',
...
```

o-mdr commented:
Hi @legeana and @jpakkane 

Thank you for looking into this. I've now made a self-contained example, could you please check out [this branch](https://github.com/o-mdr/zstd/tree/1.4.5/example) (files are also in this PR) and see if it works on your system? 

@jpakkane to your points:

`lib/compress/hist.c` is actually defined in the [original meson.build](https://github.com/facebook/zstd/blob/b951ad20a2af45a70f40caa8015a0be728eba463/build/meson/lib/meson.build#L27), it was also there in the patch

I don't see much of a difference between this three - this won't save visual space in the codebase. Is this because of the runtime vs compile-time resolution?
```
join_paths(common_dir, 'entropy_common.c')
'common/entropy_common.c'
common_dir / 'entropy_common.c'
```
I think it will make it more difficult to spot any changes between the original meson build file and a patch. I suggest we keep the original to save time on tracing the differences unless there is strong evidence in favour of a change. And if the evidence is strong, it should perhaps be committed to the zstd repo first and then replicated here to minimise diffs. 

Hope this makes it a bit easier to trace why there are linker errors. If you still experience errors, could you please post more logs?

o-mdr commented:
Hi @legeana and @jpakkane,

May be of interest to @lzutao. 

This is now ready for a review. In the process of working on this PR @lzutao and I discovered zstd isn't tested on Windows MSVC19 compiler, and it stopped working. I've updated PR and readme files to contain this information. 

Would you be able to test this on your systems when you have a free moment? Let me know if you'd like any improvements. Many thanks.

jpakkane commented:
Note that WrapDB project may **not** have any functionality patches. Only build system files. This is a matter of documented policy. If upstream is broken (for whatever reason) we will not hide it with patches. Everything in the `patch` directory has to go.

Using the `meson.build` file from the root dir does not work on macOS as it fails with link errors:

```
  "_ZSTD_selectEncodingType", referenced from:
      _ZSTD_compressSequences_internal in lib_compress_zstd_compress.c.o
```

o-mdr commented:
Hi @jpakkane could you please check if the example works? I tested on Mac OS and it works on my system. Did you copy all the files in the patch to the right places (not just a `meson.build` and `meson_options.txt`)?

legeana commented:
Readme on the branches is not really visible all that much, do we actually want to maintain it?
Also Jussi's comment is correct, patch is undesired.

jpakkane commented:
If one copies all the files then it does build correctly.

legeana commented:
The root directory doesn't have `meson.build` though.

o-mdr commented:
I guess, if you could point me to a link where patch requirements are documented I could refactor this PR (obviously I read https://mesonbuild.com/Adding-new-projects-to-wrapdb.html) or if you just want to give me a bullet points of what you would like this PR to contain I am happy to update. I feel like this work is progress, but there's certain resistence perhaps due to some guidelines that I am not aware of?

I made a `patch/zstd-1.4.5` folder to demonstrate what the patch should contain. This is purely because it was failing on testers/mainteiners side, but was working on my side. Is this because a patch is expected to only contain 1 or 2 files (meson.build and meson_options.txt)? My patch was taking from zstd upstream with minor modifications. IMHO this allows to easily track any changes in the zstd project and it's meson files. In zstd, there are several meson.build files located in several folders (so not a single build file). I wonder if this is the source of confusion, and if so, how would you like to approach this? Me bundling up all the build files for the patch in 1 larger file, or updating docs or something else.

`example` folder contains a trivial cpp example, using zipped up patch from `patch/zstd-1.4.5` to demonstrate compilation (apart from Windows+msvc) and usage of the patch. I made it because it was working on my end, but not on yours.

jpakkane commented:
The patch file can contain as many files as needed, but only build system files. This means one or more `meson.build` files, an optional `meson_options.txt` and possibly some necessary header template files, i.e. `config.h.meson`. For a typical project you'd have something like:

```
meson.build
meson_options.txt
lib/meson.build
tests/meson.build
```

These files are copied over the extracted upstream tarball. Zstd is a bit unfortunate in this case, because they don't have their `meson.build` file at the top level directory (which is required for subprojects to work) but instead it is in some subfolder. If upstream would fix that, everything would become a lot simpler. The wrapdb entry would require no patches, only a reference to upstream.

o-mdr committed:
patch to support zstd-1.4.5

o-mdr commented:
Updated to only contain patch files. Also cleaned up commit history.

eli-schwartz reviewed (changes requested):
Could you add `tests/meson.build` in order to check that it builds correctly? This is the correct (upstream) way to demonstrate compilation rather than creating your own examples.

o-mdr commented:
Hi @eli-schwartz, sure, any code examples or docs for this? I've looked into [spdlog](https://github.com/mesonbuild/spdlog/blob/1.8.0/tests/meson.build) and I get the gist. I am unsure if there are tools for pulling source code, applying a patch, and building an exe.

eli-schwartz commented:
Well in this case upstream has https://github.com/facebook/zstd/blob/dev/build/meson/tests/meson.build wired up to interact with https://github.com/facebook/zstd/tree/dev/tests

So ideally you'd be able to just use this the same way you used some other parts of their build/meson/ directory.

o-mdr committed:
removed build file that's not required

o-mdr committed:
added support for programs tests and contrib dirs

o-mdr commented:
@eli-schwartz I've removed unnecessary `build/meson/lib/meson.build` file, making this PR smaller. Also updated a patch `meson.build` file with tests, programs and contrib dirs. This allows us to compile and run tests. I've specified how to run tests in the PR description.

eli-schwartz committed:
fixup newlines

eli-schwartz commented:
Great. One last thing or two:
- files created with no eol? hmm...
- Should use https://mesonbuild.com/Wrap-dependency-system-manual.html#provide-section to auto-declare the pkg-config provided dependency.

eli-schwartz committed:
wrap should provide libzstd

o-mdr commented:
@eli-schwartz good points. Thanks for updating and merging.
